### PR TITLE
TosaToLinalg: Fix unsigned tosa.clamp

### PR DESCRIPTION
**Pull request by mgehre-amd**
Lowering tosa.clamp with unsigned integer element types needs to use `cmpi ult` instead of `cmpi slt`,
and the lower/upper bound must be interpreted as unsigned when converting to to arith.constants.

### CODE REVIEW

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp [hunk 1]
```diff
@@ -409,10 +409,15 @@ static Value createLinalgBodyCalculationForElementwiseOp(
         cast<IntegerAttr>(op->getAttr("max_int")).getValue().getSExtValue();
 
     if (intTy.isUnsignedInteger()) {
+      if (intTy.getIntOrFloatBitWidth() > 63) {
+        (void)rewriter.notifyMatchFailure(
+            op, "support for larger integers it not implemented");
```
Comment (cferry-AMD) [inline mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:414]:
```suggestion
            op, "support for 64-bit or larger integers is not implemented");
```

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp [hunk 2]
```diff
@@ -409,10 +409,15 @@ static Value createLinalgBodyCalculationForElementwiseOp(
...
+        return {};
+      }
       min = std::max(min, (int64_t)0);
-      max = std::min(
-          max,
-          APInt::getMaxValue(intTy.getIntOrFloatBitWidth()).getSExtValue());
+      max = std::min(max,
+                     (int64_t)APInt::getMaxValue(intTy.getIntOrFloatBitWidth())
+                         .getZExtValue());
```
Comment (cferry-AMD) [inline mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:420]: Good catch.

**Comment by cferry-AMD** [comment]
We should add a failing test with `ui64`:

``` 
  %failure = tosa.clamp %arg1 {min_int = -1 : i64, max_int = 5 : i64, min_fp = 1.0 : f32, max_fp = 5.0 : f32} : (tensor<1xui64>) -> tensor<1xui64>

```

Such a test will be useful to point out there is a hardcoded 63-bit limit in this pass, to be dealt with whenever larger integer support comes in.

**Review by cferry-AMD — APPROVED** [review]
Looks good!

**Review by TinaAMD — APPROVED** [review]
LGTM

One conceptual issue with clamp in general which we could think about fixing:
The [tosa spec](https://www.mlplatform.org/tosa/tosa_spec.html#_clamp) defines that the input+output type as well as the type of the attributes (min/max) need to be the same. It would be neat to enforce this (table gen + verifier), as it would prevent people from writing clip ranges which do not make sense for the given data type (as the case tested here, with unsigned inputs and a -1 lower bound).

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp (hunk 1 above)
Comment (mgehre-amd) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp]: done

**Comment by mgehre-amd** [comment]
> We should add a failing test with `ui64`:
> 
> ```
>   %failure = tosa.clamp %arg1 {min_int = -1 : i64, max_int = 5 : i64, min_fp = 1.0 : f32, max_fp = 5.0 : f32} : (tensor<1xui64>) -> tensor<1xui64>
> ```
> 
> Such a test will be useful to point out there is a hardcoded 63-bit limit in this pass, to be dealt with whenever larger integer support comes in.

Thanks! I add such test now.

**Pull request merged**